### PR TITLE
retry on franz consumer group session errors

### PR DESCRIPTION
**Pull request by samirketema**
### Description

Implements retry logic for the kafka consumer on group join and leader errors.

### Quick checks:

- [x] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [x] There is no other [pull request](https://github.com/ConduitIO/conduit-connector-kafka/pulls) for the same update/change.
- [x] I have written unit tests.
- [x] I have made sure that the PR is of reasonable size and can be easily reviewed.

## Manual Test

Steps to manually test:
1. Spin up a kafka cluster using this docker compose:
```
version: '2'
services:
  zookeeper:
    image: wurstmeister/zookeeper
    ports:
      - "2181:2181"
  kafka1:
    image: wurstmeister/kafka
    ports:
      - "9092:9092"
    environment:
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: INSIDE:PLAINTEXT
      KAFKA_LISTENERS: INSIDE://:9092
      KAFKA_ADVERTISED_LISTENERS: INSIDE://<your-local-ip>:9092
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_INTER_BROKER_LISTENER_NAME: INSIDE
    depends_on:
      - zookeeper
  kafka2:
    image: wurstmeister/kafka
    ports:
      - "9093:9093"
    environment:
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: INSIDE:PLAINTEXT
      KAFKA_LISTENERS: INSIDE://:9093
      KAFKA_ADVERTISED_LISTENERS: INSIDE://<your-local-ip>:9093
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_INTER_BROKER_LISTENER_NAME: INSIDE
    depends_on:
      - zookeeper
  kafka3:
    image: wurstmeister/kafka
    ports:
      - "9094:9094"
    environment:
      KAFKA_LISTENER_SECURITY_PROTOCOL_MAP: INSIDE:PLAINTEXT
      KAFKA_LISTENERS: INSIDE://:9094
      KAFKA_ADVERTISED_LISTENERS: INSIDE://<your-local-ip>:9094
      KAFKA_ZOOKEEPER_CONNECT: zookeeper:2181
      KAFKA_INTER_BROKER_LISTENER_NAME: INSIDE
    depends_on:
      - zookeeper

```
2. Run the following pipelines against the kafka cluster:
```yaml
version: "2.2"
pipelines:
  - id: gen-kafka
    status: running
    connectors:
      - id: gen.in
        type: source
        plugin: builtin:generator
        settings:
          format.type: "structured"
          format.options.id: "int"
          format.options.name: "string"
          format.options.company: "string"
          format.options.trial: "bool"
          rate: 0.5
          recordCount: "1000"
      - id: kafka.out
        type: destination
        plugin: standalone:kafka
        settings:
          batchBytes: "10485760"
          sdk.batch.delay: 1s
          sdk.batch.size: "50000"
          servers: <your-local-ip>:9092,<your-local-ip>:9093,<your-local-ip>:9094
          topic: test
  - id: kafka-log
    status: running
    connectors:  
      - id: kafka.in
        type: source
        plugin: standalone:kafka
        settings:
          readFromBeginning: "true"
          servers: <your-local-ip>:9092,<your-local-ip>:9093,<your-local-ip>:9094
          topic: test
        processors:
          - id: unwrapopencdc
            type: builtin:unwrap.opencdc
            plugin: ""
            condition: ""
            settings:
              field: .Payload.After
      - id: log.out
        type: destination
        plugin: builtin:log

```
3. Simulate a node going down by pausing or stopping one of the kafka node containers:

### Before

Note that the pipeline errors out and stops, preventing a recovery from the issue:
```
2024-07-10T20:49:31+00:00 ERR pipeline stopped error="node kafka-log:kafka.in stopped with error: source stream was stopped unexpectedly: error reading from source: read plugin error: failed getting a record: unable to join group session: unable to dial: dial tcp 127.0.0.1:9092: connect: connection refused" component=pipeline.Service pipeline_id=kafka-log stack=[{"file":"/Users/samir/go/src/conduit/pkg/pipeline/lifecycle.go","func":"github.com/conduitio/conduit/pkg/pipeline.(*Service).runPipeline.func2","line":582},{"file":"/Users/samir/go/src/conduit/pkg/pipeline/stream/source.go","func":"github.com/conduitio/conduit/pkg/pipeline/stream.(*SourceNode).Run","line":140},{"file":"/Users/samir/go/src/conduit/pkg/pipeline/stream/source.go","func":"github.com/conduitio/conduit/pkg/pipeline/stream.(*SourceNode).Run.func1","line":87}]
```

### After

Note that the errors are logged, but are continually retried with `ErrBackoffRetry`. This way, it's not silent, but we still can show the error.

After the broker comes back up, we can see that the connector continues processing records.

```bash
2024-07-10T20:08:17+00:00 ERR group session error, retrying: unable to join group session: unable to dial: dial tcp 127.0.0.1:9093: connect: connection refused component=plugin.standalone connector_id=kafka-log:kafka.in plugin_name=conduit-connector-kafka
2024-07-10T20:08:17+00:00 ERR join and sync loop errored component=plugin.standalone connector_id=kafka-log:kafka.in group=backoff plugin_name=conduit-connector-kafka
2024-07-10T20:08:17+00:00 WRN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=plugin.standalone connector_id=kafka-log:kafka.in plugin_name=conduit-connector-kafka req=err
2024-07-10T20:08:17+00:00 WRN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=plugin.standalone connector_id=gen-kafka:kafka.out plugin_name=conduit-connector-kafka req=err
2024-07-10T20:08:17+00:00 WRN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=plugin.standalone connector_id=kafka-log:kafka.in plugin_name=conduit-connector-kafka req=err
2024-07-10T20:08:18+00:00 WRN unable to open connection to broker addr=err component=plugin.standalone connector_id=kafka-log:kafka.in plugin_name=conduit-connector-kafka
2024-07-10T20:08:18+00:00 WRN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=plugin.standalone connector_id=gen-kafka:kafka.out plugin_name=conduit-connector-kafka req=err
2024-07-10T20:08:18+00:00 WRN produce partition load error, bumping error count on first stored batch broker=will_fail component=plugin.standalone connector_id=gen-kafka:kafka.out plugin_name=conduit-connector-kafka
2024-07-10T20:08:19+00:00 WRN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=plugin.standalone connector_id=kafka-log:kafka.in plugin_name=conduit-connector-kafka req=err
2024-07-10T20:08:19+00:00 WRN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=plugin.standalone connector_id=kafka-log:kafka.in plugin_name=conduit-connector-kafka req=err
2024-07-10T20:08:20+00:00 WRN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=plugin.standalone connector_id=kafka-log:kafka.in plugin_name=conduit-connector-kafka req=err
2024-07-10T20:08:21+00:00 WRN produce partition load error, bumping error count on first stored batch broker=will_fail component=plugin.standalone connector_id=gen-kafka:kafka.out plugin_name=conduit-connector-kafka
2024-07-10T20:08:21+00:00 WRN read from broker errored, killing connection after 0 successful responses (is SASL missing?) component=plugin.standalone connector_id=kafka-log:kafka.in plugin_name=conduit-connector-kafka req=err
2024-07-10T20:08:26+00:00 INF component=plugin connector_id=kafka-log:log.out plugin_name=builtin:log plugin_type=destination record={"key":"c2FjY2hhcmlmeQ==","metadata":{"conduit.source.connector.id":"gen-kafka:gen.in","opencdc.readAt":"1720667249647356000"},"operation":"create","payload":{"after":{"company":"redipper","id":1592616014876622300,"name":"oenanthaldehyde","trial":false},"before":null},"position":"eyJHcm91cElEIjoiOWIyNzFkN2UtYTBmNC00MzlhLTk4YWMtN2YyMjk2NjIxOTdhIiwiVG9waWMiOiJzYW1pci10ZXN0LTIiLCJQYXJ0aXRpb24iOjAsIk9mZnNldCI6MH0="}
```

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -43,6 +43,8 @@ A source is getting associated with a consumer group ID the first time the `Read
...
 | `saslUsername`       | SASL username. If provided, a password needs to be provided too.                                                                                                                                             | false    |                           |
 | `saslPassword`       | SASL password. If provided, a username needs to be provided too.                                                                                                                                             | false    |                           |
+| `RetryGroupJoinErrors`       | determines whether the connector will continually retry on group join errors                                                                                                                                              | false    | true |
+| `RetryLeaderErrors`       |    determines whether the connector will continually retry on leader errors                                                                                       | false    | true |
```
Comment (lovromazgon) [inline README.md:47]: Watch out, the option starts with a lowercase letter, and I propose formatting the default values using backticks.
```suggestion
| `retryGroupJoinErrors`       | determines whether the connector will continually retry on group join errors                                                                                                                                              | false    | `true` |
| `retryLeaderErrors`       |    determines whether the connector will continually retry on leader errors                                                                                       | false    | `true` |
```

File: common/config.go [hunk 2]
```diff
@@ -78,6 +76,7 @@ func (c Config) TryDial(ctx context.Context) error {
 		case <-ctx.Done():
 			return err
 		case <-time.After(time.Second):
+			sdk.Logger(ctx).Info().Msg("failed to dial broker, trying again...")
```
Comment (simonl2002) [inline common/config.go:79]: should this be a warning??

File: source.go [hunk 3]
```diff
@@ -90,6 +94,22 @@ func (s *Source) Open(ctx context.Context, sdkPos opencdc.Position) error {
...
+		if s.config.RetryLeaderErrors {
+			switch err {
+			case kerr.LeaderNotAvailable,
+				kerr.EligibleLeadersNotAvailable,
+				kerr.PreferredLeaderNotAvailable,
+				kerr.BrokerNotAvailable:
+				sdk.Logger(ctx).Error().Msgf("ephemeral error connecting to broker or leader, retrying: %s", err.Error())
+				return opencdc.Record{}, sdk.ErrBackoffRetry
+			}
+		}
```
[the remark this answers is not in the document]
Comment (samirketema) [reply source.go:112]: You're right - I did have this error returned locally, but I was testing out different configurations, including `kgo.KeepRetryableFetchErrors` as you mentioned. Without this config, we don't have this error returned

File: source.go [hunk 4]
```diff
@@ -90,6 +94,22 @@ func (s *Source) Open(ctx context.Context, sdkPos opencdc.Position) error {
 func (s *Source) Read(ctx context.Context) (opencdc.Record, error) {
 	rec, err := s.consumer.Consume(ctx)
 	if err != nil {
+		var errGroupSession *kgo.ErrGroupSession
+		if s.config.RetryGroupJoinErrors &&
+			(errors.As(err, &errGroupSession) || strings.Contains(err.Error(), "unable to join group session")) {
+			sdk.Logger(ctx).Error().Msgf("group session error, retrying: %s", err.Error())
+			return opencdc.Record{}, sdk.ErrBackoffRetry
+		}
```
[the remark this answers is not in the document]
Comment (samirketema) [reply source.go:102]: good point - I'll test this out and get back you to. I agree, it is a bit of a dangerous change to make, even if temporary - I'll let you know what I find testing this config out locally.

[the remark this answers is not in the document]
Comment (samirketema) [reply source.go]: ok, I decided to move the check into the franz consumer, and will push up a branch soon with some experimental code for `BlockRebalanceOnPoll`. 🙇🏾

**Comment by samirketema** [comment]
@lovromazgon ok, feedback has been addressed - thank you 🙇🏾

**Review by lovromazgon — APPROVED** [review]
👏

**Pull request merged**